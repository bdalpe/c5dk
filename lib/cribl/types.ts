export type Constructor<T, Arguments extends unknown[] = any[]> = new(...arguments_: Arguments) => T;
export type AbstractConstructor<T, Arguments extends unknown[] = any[]> = abstract new(...arguments_: Arguments) => T;

export type VariableType = 'expression' | 'number' | 'object' | 'array';

export interface VariableArguments {
	type: VariableType;
	name: string;
}

