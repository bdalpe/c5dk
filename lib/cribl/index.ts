import {Construct} from "constructs";

export interface CriblProps {
	/**
	 * The directory to output the Cribl configuration YAML manifests.
	 *
	 * @default - `CRIBL_OUTDIR`, if defined, otherwise `dist`
	 */
	outdir?: string;
}

/**
 * The base construct for all Cribl configuration.
 *
 * @example
 * ```typescript
 * const cribl = new Cribl();
 *
 * // add your config here using cribl as the
 * // const group = new Group(cribl, 'default', {});
 * // ...
 *
 * // call synth() to write the configuration to disk
 * cribl.synth();
 * ```
 */
export class Cribl extends Construct {
    public readonly outdir: string;

	constructor(props?: CriblProps) {
		super(undefined as any, '');
		this.outdir = props?.outdir ?? process.env.CRIBL_OUTDIR ?? 'dist';
	}
}

export * from "./objects";
