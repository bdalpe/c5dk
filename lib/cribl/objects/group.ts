import {Construct} from "constructs";
import {BaseConstruct} from "../base";

export interface GroupProps {}

export class Group extends BaseConstruct {
	constructor(scope: Construct, id: string, props: GroupProps) {
		super(scope, id, props);
	}
}
