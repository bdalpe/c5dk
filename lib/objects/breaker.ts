import {Construct} from "constructs";
import {ConfigConstruct} from "../config";
import Registry from "../registry";

export interface EventBreakerRulesetProps {
	lib: string;
	description?: string;
	tags?: string;
}

export class EventBreakerRuleset extends ConfigConstruct {
	constructor(scope: Construct, id: string, props: EventBreakerRulesetProps) {
		super(scope, id, props);
	}

	get kind() {
		return 'breakers';
	}

	get config(): Record<string, unknown> {
		return {...super.config, rules: this.node.children.filter(x => x instanceof EventBreakerRule).map(rule => rule.config)};
	}

	static dump(config: ConfigConstruct[]): Record<string, unknown> {
		return Object.fromEntries(config.map(c => [c.node.id, c.config]));
	}
}

Registry.register('breakers', EventBreakerRuleset);

export interface EventBreakerRuleProps {
	name: string;
	disabled: boolean;
	condition: string;
	type: string;
}

export class EventBreakerRule extends ConfigConstruct {
	constructor(scope: Construct, id: string, props: EventBreakerRuleProps) {
		// Rules can only be children of a ruleset
		if (!(scope instanceof EventBreakerRuleset)) {
			throw new Error('EventBreakerRule must be a child of EventBreakerRuleset');
		}

		super(scope, id, props);
	}

	get kind() {
		return 'rule';
	}
}
