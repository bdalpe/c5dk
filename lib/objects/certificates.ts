import {ConfigConstruct} from "../config";
import {pick} from "es-toolkit";
import {Construct} from "constructs";
import {File} from "../private/fs";
import Registry from "../registry";

export interface CertificateProps {
	/**
	 * Brief description of the certificate.
	 *
	 * Optional.
	 */
	description?: string;

	/**
	 * The certificate in PEM/Base64 format.
	 */
	certificate: string;

	/**
	 * The private key in PEM/Base64 format.
	 */
	privateKey: string;

	/**
	 * The passphrase for the private key.
	 *
	 * Optional.
	 */
	passphrase?: string;

	/**
	 * The CA certificate in PEM/Base64 format.
	 *
	 * Optional.
	 */
	caCert?: string;
}

export class Certificate extends ConfigConstruct {
	protected certificate: string;
	protected privateKey: string;
	protected passphrase : string | undefined;
	protected caCert: string | undefined;

	constructor(scope: Construct, id: string, props: CertificateProps) {
		super(scope, id, pick(props, ['description']));

		this.certificate = props.certificate;
		this.privateKey = props.privateKey;
		this.passphrase = props.passphrase;
		this.caCert = props.caCert;
	}

	get kind() {
		return 'certificates';
	}

	synth(): void {
		new File(this.path('local', 'cribl', 'auth', 'certs', `${this.node.id}.pem`)).write(this.certificate);
		new File(this.path('local', 'cribl', 'auth', 'certs', `${this.node.id}.key`)).write(this.privateKey);
		if (this.passphrase) new File(this.path('local', 'cribl', 'auth', 'certs', `${this.node.id}.pas`)).write(this.passphrase);
		if (this.caCert) new File(this.path('local', 'cribl', 'auth', 'certs', `${this.node.id}.crt`)).write(this.caCert);
	}

	static dump(config: ConfigConstruct[]): Record<string, unknown> {
		return Object.fromEntries(config.map(c => [c.node.id, c.config]));
	}
}

Registry.register('certificates', Certificate);
