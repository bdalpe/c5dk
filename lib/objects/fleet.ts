import {ContainerConstruct} from "../container";

export interface GroupProps {}

export class Fleet extends ContainerConstruct {
	_package = 'edge';

	get kind() {
		return 'group';
	}
}
