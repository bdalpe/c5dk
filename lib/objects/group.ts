import {ContainerConstruct} from "../container";
import {Construct} from "constructs";
import Registry from "../registry";

export interface GroupProps {
	description?: string;
}

export class WorkerGroup extends ContainerConstruct {
	constructor(scope: Construct, id: string, protected props?: GroupProps) {
		super(scope, id, props ?? {});
	}

	get kind() {
		return 'groups';
	}

	static dump(config: ContainerConstruct[]): Record<string, unknown> | undefined {
		return Object.fromEntries(config.map(c => [c.node.id, c.config]));
	}
}

Registry.registerContainer('groups', WorkerGroup)
