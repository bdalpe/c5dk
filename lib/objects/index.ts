export * from './output';
export * from './breaker';
export * from './certificates';
export * from './fleet';
export * from './license';
export * from './sample';
export * from './scripts';
export * from './variable';
export * from './regex'
