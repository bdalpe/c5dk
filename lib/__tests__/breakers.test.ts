import {describe, it, expect, vi, beforeEach} from 'vitest';
import {fs, vol} from 'memfs'
import {Cribl} from "../cribl";
import {EventBreakerRule, EventBreakerRuleset} from "../cribl/objects/breaker";

vi.mock('node:fs');
vi.mock('node:fs/promises');

let cribl: Cribl;

beforeEach(() => {
	// reset the state of in-memory fs
	vol.reset()

	cribl = new Cribl({outdir: '/tmp'});
})

describe('Event Breakers', () => {
	describe('Rulesets', () => {
		it('should be written to a file', () => {
			new EventBreakerRuleset(cribl, 'test', {lib: 'custom'});
			cribl.synth();

			const volume = vol.toJSON();
			expect(Object.keys(volume)).toEqual(['/tmp/local/cribl/breakers.yml']);
			expect(volume['/tmp/local/cribl/breakers.yml']).toMatchSnapshot();
		});

		describe('with Rules', () => {
			it('should be written to a file', () => {
				const ruleset = new EventBreakerRuleset(cribl, 'test', {lib: 'custom'});
				new EventBreakerRule(ruleset, 'first', {name: 'first', disabled: false, condition: 'true', type: 'test'});

				cribl.synth();

				const volume = vol.toJSON();
				expect(Object.keys(volume)).toEqual(['/tmp/local/cribl/breakers.yml']);
				expect(volume['/tmp/local/cribl/breakers.yml']).toMatchSnapshot();
			});
		});
	});
});
