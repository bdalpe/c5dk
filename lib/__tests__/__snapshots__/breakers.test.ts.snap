// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Event Breakers > Rulesets > should be written to a file 1`] = `
"test:
  lib: custom
  rules: []
"
`;

exports[`Event Breakers > Rulesets > with Rules > should be written to a file 1`] = `
"test:
  lib: custom
  rules:
    - name: first
      disabled: false
      condition: 'true'
      type: test
"
`;
